### PR TITLE
feat(import): add MLFlow import support (kit import mlflow://)

### LINKED ISSUE #1014: Implement MLFlow import: `kit import mlflow://` for experiment-to-ModelKit workflow
Data scientists train models using MLFlow but must manually package them as ModelKits for deployment. This breaks the workflow continuity and creates opportunities for version mismatches between what was tracked in MLFlow and what gets deployed.

**Proposed Solution**

Add MLFlow as an import source with URI syntax: `kit import mlflow://[tracking_uri/]experiments/{exp_id}/runs/{run_id}`

### Architecture

Leverage existing `kit init` pipeline (which already runs during import). Your work is:

1. **MLFlow URI handler** - parse URIs, extract tracking server + run identifiers  
2. **Artifact downloader** - use pull run artifacts to a temp directory
3. **Metadata enrichment** - inject MLFlow provenance into generated Kitfile
4. **Error handling** - deal with auth, large files, incomplete runs

### Implementation Details

**URI Format**
```bash
# With explicit tracking URI
kit import mlflow://mlflow.company.com/experiments/42/runs/abc123 -t mymodel:v1

# Using MLFLOW_TRACKING_URI env var
export MLFLOW_TRACKING_URI=http://localhost:5000
kit import mlflow://experiments/42/runs/abc123 -t mymodel:v1

# Short form (uses default experiment)
kit import mlflow://runs/abc123 -t mymodel:v1
```

**Data Flow**
```
1. Parse mlflow:// URI → tracking_uri, experiment_id, run_id
2. MLFlow client: fetch run metadata + list artifacts
3. Download artifacts to temp dir (filtered by size/type)
4. Run kit init on temp dir → generates Kitfile
5. Augment Kitfile with MLFlow provenance metadata
6. Pack ModelKit using existing pipeline
```
### Implementation Challenges

**1. Authentication Hell**

MLFlow supports multiple auth mechanisms with no standard:
- Basic auth (username/password)
- Token-based (custom headers)
- Cloud provider auth (AWS IAM, GCP service accounts) for artifact stores
- No auth (local/trusted network)

**Approach**: 
- Support `MLFLOW_TRACKING_URI` and `MLFLOW_TRACKING_TOKEN` env vars
- Document that artifact store auth must be handled separately (AWS credentials, GCS keys, etc.)
- Fail fast with clear error messages when auth is missing

**2. Large Artifact Handling**

50GB model checkpoint will timeout/OOM with naive download make sure to use filters to download only what needs to be packed

**3. Incomplete/Failed Runs**

MLFlow runs can be RUNNING, FAILED, or KILLED. Artifacts may be partial.  Only import FINISHED runs

**4. Storage Backend Diversity**

MLFlow artifact stores can be:
- Local filesystem (`file:///`)
- S3 (`s3://`)
- GCS (`gs://`)
- Azure (`wasbs://`)
- SFTP, NFS, etc.

We can either implement these backends on Kit or rely on MLFlow client.

### PR DESCRIPTION
**Pull request by akagami-harsh**
### Description
Adds MLFlow as an import source for `kit import`, us to package MLFlow experiment runs as ModelKits in a single command. Supports three URI forms: `mlflow://host/experiments/{id}/runs/{id}`, `mlflow://experiments/{id}/runs/{id}`, and the short `mlflow://runs/{id}`. The tracking server is resolved from the URI host, the `MLFLOW_TRACKING_URI` env var, or defaults to `http://localhost:5000`. Authentication is handled via `--token` flag or `MLFLOW_TRACKING_TOKEN` env var. Only `FINISHED` runs can be imported `RUNNING`, `FAILED`, and `KILLED` runs are rejected with a clear error. Artifacts are downloaded concurrently through the tracking server's artifact proxy, a Kitfile is auto-generated (or a user-provided one is used), and MLFlow provenance metadata (run ID, experiment ID, metrics, params, timestamps) is injected into `model.parameters` before packing.

Resolves #1014

### Linked issues
- https://github.com/kitops-ml/kitops/issues/1014

### AI-Assisted Code

- [x] This PR contains AI-generated code that I have reviewed and tested
- [x] I take full responsibility for all code in this PR, regardless of how it was created

### CODE REVIEW

**Comment by akagami-harsh** [comment]
tested on local setup:

- Start MLFlow locally
```bash
(venv) harshvir@msi:~/kitops$ mlflow server --host 0.0.0.0 --port 5000 
Backend store URI not provided. Using sqlite:///mlflow.db
Registry store URI not provided. Using backend store URI.
[MLflow] Security middleware enabled with default settings (localhost-only). To allow connections from other hosts, use --host 0.0.0.0 and configure --allowed-hosts and --cors-allowed-origins.
INFO:     Uvicorn running on http://0.0.0.0:5000 (Press CTRL+C to quit)
INFO:     Started parent process [82500]
2026/03/17 15:41:36 INFO mlflow.server.jobs.utils: Starting huey consumer for job function invoke_scorer
2026/03/17 15:41:36 INFO mlflow.server.jobs.utils: Starting huey consumer for job function run_online_trace_scorer
```
- Log a quick test run
```bash
(venv) harshvir@msi:~/kitops$ python3 testrun.py 
Run ID: 2de0dbba9b164eee9a6a27551afb6502
🏃 View run test-model at: http://localhost:5000/#/experiments/0/runs/2de0dbba9b164eee9a6a27551afb6502
🧪 View experiment at: http://localhost:5000/#/experiments/0
```

- Import the run using above run id
```bash
(venv) harshvir@msi:~/kitops$ ./kit import mlflow://runs/2de0dbba9b164eee9a6a27551afb6502 -t my-test-raw
:v1
Connecting to MLFlow tracking server http://localhost:5000
Found run "test-model" (experiment 0) — status FINISHED
Found 1 artifact file(s) for run 2de0dbba9b164eee9a6a27551afb6502
Downloading artifact model.pkl
Generated Kitfile:

manifestVersion: 1.0.0
package:
  name: test-model
  authors: [mlflow-exp0]
model:
  name: model
  path: model.pkl

Would you like to edit Kitfile before packing? (y/N): 
Updated Kitfile with MLFlow provenance:

manifestVersion: 1.0.0
package:
  name: test-model
  version: "20260317082144"
  description: Imported from MLFlow run 2de0dbba9b164eee9a6a27551afb6502 — (test-model) — tracking server http://localhost:5000
  authors: [mlflow-exp0]
model:
  name: model
  path: model.pkl
  parameters:
    mlflow_provenance:
      mlflow_end_time: "2026-03-17T08:21:44Z"
      mlflow_experiment_id: "0"
      mlflow_metrics:
        accuracy: 0.95
      mlflow_params:
        learning_rate: "0.01"
      mlflow_run_id: 2de0dbba9b164eee9a6a27551afb6502
      mlflow_run_name: test-model
      mlflow_run_status: FINISHED
      mlflow_start_time: "2026-03-17T08:21:44Z"
      mlflow_tracking_uri: http://localhost:5000

Packing model to my-test-raw:v1
Already saved model layer: sha256:81a20b25d26d08405a4cfb7e10862c400cb1792cff58d08b928c3a3428fbf77c
Saved configuration: sha256:d2437f172bf3eff84f8cbc29a14dd672e22cde50593bd139358f6be5f80943f3
Saved manifest to storage: sha256:771f1f22f20fcbdbe07a4604544d7d02bdebc72dd2f4f079aa80c19dfa8045df
Model is packed as my-test-raw:v1
```


- Check it was packed
```bash
(venv) harshvir@msi:~/kitops$ ./kit list
REPOSITORY      TAG      MAINTAINER    NAME         SIZE      DIGEST
model           v1       mlflow-exp0   test-model   2.0 KiB   sha256:9a8d2e31045803d283e382895d9689dce663071b27e8ac7c4d9803e45cd15b54
my-test-model   v5       mlflow-exp0   test-model   2.0 KiB   sha256:9a8d2e31045803d283e382895d9689dce663071b27e8ac7c4d9803e45cd15b54
my-test-model   <none>   mlflow-exp0   test-model   2.0 KiB   sha256:d72828c4ff7377e2eaf22ae030f3d5dd93a1b355ae87d9326f26d3f9deca1e48
my-test-model   v1       mlflow-exp0   test-model   2.0 KiB   sha256:e63c69003ab40166f64cf286610bfc57db1f83b71cdc18e845177073cacf18da
my-test-raw     v1       mlflow-exp0   test-model   2.0 KiB   sha256:771f1f22f20fcbdbe07a4604544d7d02bdebc72dd2f4f079aa80c19dfa8045df
```
```
(venv) harshvir@msi:~/kitops$ ./kit inspect my-test-raw:v1
{
  "digest": "sha256:771f1f22f20fcbdbe07a4604544d7d02bdebc72dd2f4f079aa80c19dfa8045df",
  "cliVersion": "unknown",
  "kitfile": {
    "manifestVersion": "1.0.0",
    "package": {
      "name": "test-model",
      "version": "20260317082144",
      "description": "Imported from MLFlow run 2de0dbba9b164eee9a6a27551afb6502 — (test-model) — tracking server http://localhost:5000",
      "authors": [
        "mlflow-exp0"
      ]
    },
    "model": {
      "name": "model",
      "path": "model.pkl",
      "parameters": {
        "mlflow_provenance": {
          "mlflow_end_time": "2026-03-17T08:21:44Z",
          "mlflow_experiment_id": "0",
          "mlflow_metrics": {
            "accuracy": 0.95
          },
          "mlflow_params": {
            "learning_rate": "0.01"
          },
          "mlflow_run_id": "2de0dbba9b164eee9a6a27551afb6502",
          "mlflow_run_name": "test-model",
          "mlflow_run_status": "FINISHED",
          "mlflow_start_time": "2026-03-17T08:21:44Z",
          "mlflow_tracking_uri": "http://localhost:5000"
        }
      },
      "digest": "sha256:81a20b25d26d08405a4cfb7e10862c400cb1792cff58d08b928c3a3428fbf77c",
      "diffId": "sha256:81a20b25d26d08405a4cfb7e10862c400cb1792cff58d08b928c3a3428fbf77c"
    }
  },
  "manifest": {
    "schemaVersion": 2,
    "artifactType": "application/vnd.kitops.modelkit.manifest.v1+json",
    "config": {
      "mediaType": "application/vnd.kitops.modelkit.config.v1+json",
      "digest": "sha256:d2437f172bf3eff84f8cbc29a14dd672e22cde50593bd139358f6be5f80943f3",
      "size": 838
    },
    "layers": [
      {
        "mediaType": "application/vnd.kitops.modelkit.model.v1.tar",
        "digest": "sha256:81a20b25d26d08405a4cfb7e10862c400cb1792cff58d08b928c3a3428fbf77c",
        "size": 2048,
        "annotations": {
          "org.cncf.model.filepath": "model.pkl"
        }
      }
    ],
    "annotations": {
      "ml.kitops.modelkit.cli-version": "unknown"
    }
  }
}
```

File: pkg/lib/mlflow/download.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	progress, plog := output.NewDownloadProgress()
+
+	for _, f := range files {
+		if err := sem.Acquire(errCtx, 1); err != nil {
+			semErr = err
+			break
+		}
+
+		artifactPath := f.Path
+		destPath := filepath.Join(destDir, filepath.FromSlash(artifactPath))
```
Comment (gorkem) [inline pkg/lib/mlflow/download.go]: We should verify the path so that a comprimised MLFlow server does not write files anywhere on the filesystem by returning artifact paths containing `../ `components.

`filepath.Join` normalizes the path but does not confine it within `destDir`. For example:

**Fix:** Call `filesystem.VerifySubpath(destDir, filepath.FromSlash(artifactPath))` before constructing destPath and return an error if it fails.

File: pkg/lib/mlflow/client.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, parseAPIError(resp)
+	}
+
+	var body struct {
+		Files     []ArtifactFileInfo `json:"files"`
+		RootURI   string             `json:"root_uri"`
+		NextToken string             `json:"next_page_token"`
```
Comment (gorkem) [inline pkg/lib/mlflow/client.go]: This is never used, should it be?

File: pkg/cmd/kitimport/cmd.go
Comment (gorkem) [inline pkg/cmd/kitimport/cmd.go]: MLFlow downloads now also use this flag but the description only mentions HuggingFace, which will confuse users.

File: pkg/lib/mlflow/download.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		Name: name,
+		Path: path,
+	})
+	return &parent.Subdirs[len(parent.Subdirs)-1]
+}
+
+// splitPath splits a slash-separated artifact path into components.
+func splitPath(p string) []string {
+	var parts []string
+	for _, seg := range filepath.SplitList(p) {
```
Comment (gorkem) [inline pkg/lib/mlflow/download.go]: Feels like this code incorrectly uses `filepath.SplitList` or the comments are confusing

File: pkg/lib/mlflow/uri.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	return &ParsedURI{
+		TrackingURI:  trackingURI,
+		ExperimentID: expID,
+		RunID:        runID,
+	}, nil
+}
+
+// isPathKeyword returns true if the segment is a well-known first path segment
+// rather than a host name.
+func isPathKeyword(seg string) bool {
```
Comment (gorkem) [inline pkg/lib/mlflow/uri.go]: Is this used ?

Comment (akagami-harsh) [reply pkg/lib/mlflow/uri.go]: I was previously using this function, but I decided to inline the logic since it was quite small. I removed the unused function in  https://github.com/kitops-ml/kitops/pull/1124/commits/80ac245921aa3d29af87bb0c09469af0972a9050

File: pkg/lib/mlflow/download.go (hunk 3 above)
Comment (akagami-harsh) [reply pkg/lib/mlflow/download.go]: Simplified the the `splitPath` logic and Replaced with `strings.Split(p, "/")` in https://github.com/kitops-ml/kitops/pull/1124/commits/972209528c089375f1b32010b285840420b7ac6c

File: pkg/cmd/kitimport/cmd.go
Comment (akagami-harsh) [reply pkg/cmd/kitimport/cmd.go]: Updated the flag description to include MLFlow too in https://github.com/kitops-ml/kitops/pull/1124/commits/04df9d94669a3107abab0b9038311a9e56dc61c6

File: pkg/lib/mlflow/client.go (hunk 2 above)
Comment (akagami-harsh) [reply pkg/lib/mlflow/client.go]: No, its not being used, remove it and cleaned up some more structs in https://github.com/kitops-ml/kitops/pull/1124/commits/756491c42424144ab49cfc9f89ebbb81d63ea558

File: pkg/lib/mlflow/download.go (hunk 1 above)
Comment (akagami-harsh) [reply pkg/lib/mlflow/download.go]: Good catch! Fixed with `filesystem.VerifySubpath` as suggested [here](https://github.com/kitops-ml/kitops/pull/1124/commits/972209528c089375f1b32010b285840420b7ac6c)

**Comment by akagami-harsh** [comment]
@gorkem, Could you please take a look again
